File: test/language/sql_constraints/python/FieldConstraintsTest.py
import unittest
import os
import apsw

from testutils import getZserioApi, getApiDir

class FieldConstraintsTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "sql_constraints.zs")
        cls._fileName = os.path.join(getApiDir(os.path.dirname(__file__)), "field_constraints_test.sqlite")

    def setUp(self):
        if os.path.exists(self._fileName):
            os.remove(self._fileName)
        self._database = self.api.TestDb.fromFile(self._fileName)
        self._database.createSchema()
        self._constraintsTable = self._database.field_constraints_table

    def tearDown(self):
        self._database.close()

    def testWithoutSql(self):
        rowDict = self._createRowDict()
        rowDict['withoutSql'] = None
        self._writeRowDict(rowDict)

    def testSqlNotNull(self):
        rowDict = self._createRowDict()
        rowDict['sqlNotNull'] = None
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue(
                "NOT NULL constraint failed: fieldConstraintsTable.sqlNotNull" in str(context.exception))

    def testSqlDefaultNull(self):
        rowDict = self._createRowDict()
        rowDict['sqlDefaultNull'] = None
        self._writeRowDict(rowDict)

    def testSqlCheckConstant(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckConstant'] = self.WRONG_CONSTRAINTS_CONSTANT
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: fieldConstraintsTable" in str(context.exception))

    def testSqlCheckImportedConstant(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckImportedConstant'] = self.WRONG_IMPORTED_CONSTRAINTS_CONSTANT
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: fieldConstraintsTable" in str(context.exception))

    def testSqlCheckUnicodeEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckUnicodeEscape'] = self.WRONG_UNICODE_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: fieldConstraintsTable" in str(context.exception))

    def testSqlCheckHexEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckHexEscape'] = self.WRONG_HEX_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: fieldConstraintsTable" in str(context.exception))

    def testSqlCheckOctalEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckOctalEscape'] = self.WRONG_OCTAL_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: fieldConstraintsTable" in str(context.exception))

    def _createRowDict(self):
        return {
            'primaryKey' : 1,
            'withoutSql' : 1,
            'sqlNotNull' : 1,
            'sqlDefaultNull' : 1,
            'sqlCheckConstant' : 1,
            'sqlCheckImportedConstant' : 1,
            'sqlCheckUnicodeEscape' : self.UNICODE_ESCAPE_CONST,
            'sqlCheckHexEscape' : self.HEX_ESCAPE_CONST,
            'sqlCheckOctalEscape' : self.OCTAL_ESCAPE_CONST
        }

    def _writeRowDict(self, rowDict):
        row = [(
            rowDict['primaryKey'],
            rowDict['withoutSql'],
            rowDict['sqlNotNull'],
            rowDict['sqlDefaultNull'],
            rowDict['sqlCheckConstant'],
            rowDict['sqlCheckImportedConstant'],
            rowDict['sqlCheckUnicodeEscape'],
            rowDict['sqlCheckHexEscape'],
            rowDict['sqlCheckOctalEscape']
        )]
        self._constraintsTable.write(row)

    UNICODE_ESCAPE_CONST = 1
    HEX_ESCAPE_CONST = 2
    OCTAL_ESCAPE_CONST = 3

    WRONG_CONSTRAINTS_CONSTANT = 124
    WRONG_IMPORTED_CONSTRAINTS_CONSTANT = 322

    WRONG_UNICODE_ESCAPE_CONST = 0
    WRONG_HEX_ESCAPE_CONST = 0
    WRONG_OCTAL_ESCAPE_CONST = 0

File: test/language/sql_constraints/cpp/FieldConstraintsTest.cpp
#include <cstdio>
#include <vector>

#include "gtest/gtest.h"

#include "sql_constraints/TestDb.h"
#include "sql_constraints/field_constraints/FieldConstraintsTable.h"

#include "zserio/SqliteException.h"
#include "zserio/RebindAlloc.h"

namespace sql_constraints
{
namespace field_constraints
{

using allocator_type = FieldConstraintsTable::allocator_type;
template <typename T>
using vector_type = std::vector<T, zserio::RebindAlloc<allocator_type, T>>;

class FieldConstraintsTest : public ::testing::Test
{
public:
    FieldConstraintsTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new sql_constraints::TestDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~FieldConstraintsTest()
    {
        delete m_database;
    }

protected:
    void fillRow(FieldConstraintsTable::Row& row)
    {
        row.setWithoutSql(1);
        row.setSqlNotNull(1);
        row.setSqlDefaultNull(1);
        row.setSqlCheckConstant(1);
        row.setSqlCheckImportedConstant(1);
        row.setSqlCheckUnicodeEscape(UNICODE_ESCAPE_CONST);
        row.setSqlCheckHexEscape(HEX_ESCAPE_CONST);
        row.setSqlCheckOctalEscape(OCTAL_ESCAPE_CONST);
    }

    static const char DB_FILE_NAME[];

    static const uint8_t UNICODE_ESCAPE_CONST;
    static const uint8_t HEX_ESCAPE_CONST;
    static const uint8_t OCTAL_ESCAPE_CONST;

    static const uint16_t WRONG_CONSTRAINTS_CONSTANT;
    static const uint32_t WRONG_IMPORTED_CONSTRAINTS_CONSTANT;

    static const uint8_t WRONG_UNICODE_ESCAPE_CONST;
    static const uint8_t WRONG_HEX_ESCAPE_CONST;
    static const uint8_t WRONG_OCTAL_ESCAPE_CONST;

    sql_constraints::TestDb* m_database;
};

const char FieldConstraintsTest::DB_FILE_NAME[] =
        "language/sql_constraints/field_constraints_test.sqlite";

const uint8_t FieldConstraintsTest::UNICODE_ESCAPE_CONST = 1;
const uint8_t FieldConstraintsTest::HEX_ESCAPE_CONST = 2;
const uint8_t FieldConstraintsTest::OCTAL_ESCAPE_CONST = 3;

const uint16_t FieldConstraintsTest::WRONG_CONSTRAINTS_CONSTANT = 124;
const uint32_t FieldConstraintsTest::WRONG_IMPORTED_CONSTRAINTS_CONSTANT = 322;

const uint8_t FieldConstraintsTest::WRONG_UNICODE_ESCAPE_CONST = 0;
const uint8_t FieldConstraintsTest::WRONG_HEX_ESCAPE_CONST = 0;
const uint8_t FieldConstraintsTest::WRONG_OCTAL_ESCAPE_CONST = 0;

TEST_F(FieldConstraintsTest, withoutSql)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.resetWithoutSql();
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_NO_THROW(fieldConstraintsTable.write(rows));
}

TEST_F(FieldConstraintsTest, sqlNotNull)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.resetSqlNotNull();
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(FieldConstraintsTest, sqlDefaultNull)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.resetSqlDefaultNull();
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_NO_THROW(fieldConstraintsTable.write(rows));
}

TEST_F(FieldConstraintsTest, sqlCheckConstant)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckConstant(WRONG_CONSTRAINTS_CONSTANT);
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(FieldConstraintsTest, sqlCheckImportedConstant)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckImportedConstant(WRONG_IMPORTED_CONSTRAINTS_CONSTANT);
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(FieldConstraintsTest, sqlCheckUnicodeEscape)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckUnicodeEscape(WRONG_UNICODE_ESCAPE_CONST);
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(FieldConstraintsTest, sqlCheckHexEscape)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckHexEscape(WRONG_HEX_ESCAPE_CONST);
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(FieldConstraintsTest, sqlCheckOctalEscape)
{
    FieldConstraintsTable& fieldConstraintsTable = m_database->getFieldConstraintsTable();
    FieldConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckOctalEscape(WRONG_OCTAL_ESCAPE_CONST);
    vector_type<FieldConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(fieldConstraintsTable.write(rows), zserio::SqliteException);
}

} // namespace field_constraints
} // namespace sql_constraints
